### PR TITLE
trinity: Proxied services now record remote exceptions' tracebacks 

### PR DESCRIPTION
**Pull request by gsalgado**
Closes: #956

### CODE REVIEW

File: trinity/chains/__init__.py [hunk 1]
```diff
@@ -144,6 +148,33 @@ def initialize_database(chain_config: ChainConfig, chaindb: AsyncChainDB) -> Non
             )
 
 
+class TracebackKeeper:
```
Comment (pipermerriam) [inline trinity/chains/__init__.py]: *naming nitpick*

maybe `TracebackPreserver` or `TracebackRecorder`?

Comment (cburgdorf) [reply trinity/chains/__init__.py]: :+1:  on `TracebackRecorder`

File: trinity/chains/__init__.py [hunk 2]
```diff
@@ -144,6 +148,33 @@ def initialize_database(chain_config: ChainConfig, chaindb: AsyncChainDB) -> Non
...
+
+    def __dir__(self) -> List[str]:
+        return dir(self.obj)
+
+    def __getattr__(self, name: str) -> Any:
+        attr = getattr(self.obj, name)
+        if not inspect.ismethod(attr):
+            return attr
+
+        def wrapper(*args: Any, **kwargs: Any) -> None:
```
Comment (pipermerriam) [inline trinity/chains/__init__.py]: **maybe** lift this out of the function body to the top level of the module and then this function body would be 

```
if not inspect.ismethod(attr):
    return attr
else:
    return record_traceback_on_error(attr)
```

**Comment by gsalgado** [comment]
I've improved this a bit based on [this trick](https://bugs.python.org/issue13831)

It uses exception chaining, which means we don't need any extra hacks to special case the logging of those exceptions. This is how it looks like:

```
   ERROR  07-11 11:45:15     service  Unexpected error in <p2p.chain.FastChainSyncer object at 0x7f7fa2b3cf60>, exiting
trinity.chains.RemoteTraceback:
"""
Traceback (most recent call last):
  File "/home/salgado/src/py-evm/trinity/chains/__init__.py", line 176, in wrapper
    return attr(*args, **kwargs)
  File "/home/salgado/src/py-evm/evm/db/header.py", line 86, in get_canonical_block_hash
    validate_block_number(block_number, title="Block Number")
  File "/home/salgado/src/py-evm/evm/validation.py", line 201, in validate_block_number
    validate_is_integer(block_number, title="Block Number")
  File "/home/salgado/src/py-evm/evm/validation.py", line 37, in validate_is_integer
    "{title} must be a an integer.  Got: {0}".format(type(value), title=title)
evm.exceptions.ValidationError: Block Number must be a an integer.  Got: <class 'bytes'>
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/salgado/src/py-evm/p2p/service.py", line 55, in run
    await self._run()
  File "/home/salgado/src/py-evm/p2p/chain.py", line 120, in _run
    await self.wait(self.db.coro_get_canonical_block_hash(b'foo'))
  File "/home/salgado/src/py-evm/p2p/cancel_token.py", line 102, in wait
    return await self.wait_first(awaitable, token=token, timeout=timeout)
  File "/home/salgado/src/py-evm/p2p/cancel_token.py", line 123, in wait_first
    return await wait_with_token(*awaitables, token=token_chain, timeout=timeout)
  File "/home/salgado/src/py-evm/p2p/cancel_token.py", line 160, in wait_with_token
    return done.pop().result()
  File "/home/salgado/src/py-evm/trinity/utils/mp.py", line 27, in method
    args,
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/lib/python3.6/multiprocessing/managers.py", line 772, in _callmethod
    raise convert_to_error(kind, result)
evm.exceptions.ValidationError: Block Number must be a an integer.  Got: <class 'bytes'>
 ```

**Pull request merged**